### PR TITLE
feat: integrated openAI rust plugin 

### PR DESCRIPTION
**Pull request by yatendra2001**
### Feature Preview

1. Created flowy-plugins rust lib.
2. Integrated plugin to interface.
3. Added API call to openai gpt-3.5-turbo

**How to test?**
1. Enter your open-ai API key in settings.
2. Then tap on sidebar username.

|Demo answer for a prompt|
|---|
| Screenshot 2023-08-31 at 7 01 04 AM |

Related to #2782 

---

#### PR Checklist

- [x] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [x] I've listed at least one issue that this PR fixes in the description above.
- [x] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

**Comment by LucasXu0** [comment]
Hi, @yatendra2001. Please check the analyzed issues.

**Comment by yatendra2001** [comment]
I think the Rust CI- issues's coming from flowy-server/tests/supabase_test/folder_test.rs .
```
failures:

---- supabase_test::folder_test::supabase_create_workspace_test stdout ----
thread 'supabase_test::folder_test::supabase_create_workspace_test' panicked at 'called `Result::unwrap()` on an `Err` value: builder error: relative URL without a base

Caused by:
    relative URL without a base', flowy-server/tests/supabase_test/folder_test.rs:25:61
```

**and just fixing flutter analyzer issue right now.**

**Comment by appflowy** [comment]
Hi @yatendra2001, You can  merge the latest main branch. It should be fixed.

**Comment by yatendra2001** [comment]
> Hi @yatendra2001, You can merge the latest main branch. It should be fixed.

alright, doing it now.

File: frontend/appflowy_flutter/packages/flowy_infra_ui/example/macos/Podfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-platform :osx, '10.11'
+platform :osx, '10.14'
```
Comment (appflowy) [inline frontend/appflowy_flutter/packages/flowy_infra_ui/example/macos/Podfile:1]: Is it necessary to update this?

File: frontend/rust-lib/flowy-plugins/src/event_handler.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    .header("Authorization", format!("Bearer {}", data.open_ai_key))
+    .json(&body)
+    .send()
+    .await?
+    .json()
+    .await?;
+
+  // Extract index and content
+  let choice = &response.choices[0];
+  println!("Index: {}", choice.index);
```
Comment (appflowy) [inline frontend/rust-lib/flowy-plugins/src/event_handler.rs:62]: Please don't use print

File: frontend/appflowy_flutter/packages/flowy_infra_ui/example/macos/Podfile (hunk 1 above)
Comment (yatendra2001) [reply frontend/appflowy_flutter/packages/flowy_infra_ui/example/macos/Podfile:1]: actually, it wasn't working before updating this. showed some warning.

i'll just test it again and if it works, will change it back to `10.11`

File: frontend/rust-lib/flowy-plugins/src/event_handler.rs (hunk 2 above)
Comment (yatendra2001) [reply frontend/rust-lib/flowy-plugins/src/event_handler.rs:62]: yeah sure, I'll just remove it. Used it for testing, but forgot to remove 🥲

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart [hunk 3]
```diff
@@ -92,10 +93,26 @@ class SidebarUser extends StatelessWidget {
...
-      overflow: TextOverflow.ellipsis,
-      color: Theme.of(context).colorScheme.tertiary,
+    return GestureDetector(
+      onTap: () async {
+        final result = await OpenAIService.requestTextCompletion(
+          prompt: 'tell a story',
+          openAIKey: state.userProfile.openaiKey,
+          model: "gpt-3.5-turbo",
+        );
+        debugPrint("OPEN AI KEY:${state.userProfile.openaiKey}");
```
Comment (appflowy) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart:103]: Also remove this print. Actually, you can use tracing log.

File: frontend/rust-lib/flowy-plugins/src/event_handler.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        role: "user".to_string(),
+        content: data.prompt.to_string(),
+      },
+    ],
+  };
+
+  // Make the API call
+  let client = reqwest::Client::new();
+  let response: ApiResponse = client
+    .post("https://api.openai.com/v1/chat/completions")
```
Comment (appflowy) [inline frontend/rust-lib/flowy-plugins/src/event_handler.rs:51]: You can try this crate. https://github.com/dongri/openai-api-rs.

Comment (yatendra2001) [reply frontend/rust-lib/flowy-plugins/src/event_handler.rs:51]: yeah, i tried these but I did not use them because Openai is evolving, if they changed their endpoints and models to something else then we're dependent on the crates.
Instead, we can maintain ourselves and keep the project light by using only the required endpoints.

wdyt??

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart (hunk 3 above)
Comment (yatendra2001) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart:103]: alright i'll just do it right away.

Comment (appflowy) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart:103]: yep

File: frontend/rust-lib/flowy-plugins/src/event_handler.rs (hunk 4 above)
Comment (appflowy) [reply frontend/rust-lib/flowy-plugins/src/event_handler.rs:51]: it's ok, just keep it. But I think we will use other features of Open AI soon.

File: frontend/rust-lib/flowy-plugins/src/entities.rs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  #[pb(index = 1)]
+  pub request_id: String,
+
+  // Model: Either text-davinci-003 or gpt-3.5-turbo
+  #[pb(index = 2)]
+  pub model: String,
+
+  // Prompt to query gpt
+  #[pb(index = 3)]
+  pub prompt: String,
```
Comment (appflowy) [inline frontend/rust-lib/flowy-plugins/src/entities.rs:19]: Actually, I think there's no need to pass the `open_ai_key` and the model. We can have a settings page for OpenAI to handle this. I will update the logic here after merging this PR.

File: frontend/appflowy_flutter/packages/appflowy_backend/example/macos/Runner.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -544,7 +544,7 @@
 				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
-				MACOSX_DEPLOYMENT_TARGET = 10.11;
```
Comment (appflowy) [inline frontend/appflowy_flutter/packages/appflowy_backend/example/macos/Runner.xcodeproj/project.pbxproj:547]: please revert all these setting.

File: frontend/rust-lib/flowy-plugins/src/entities.rs (hunk 5 above)
Comment (yatendra2001) [reply frontend/rust-lib/flowy-plugins/src/entities.rs:19]: Sure, i don't need to make any changes for this part of code right ??

File: frontend/appflowy_flutter/packages/appflowy_backend/example/macos/Runner.xcodeproj/project.pbxproj (hunk 6 above)
Comment (yatendra2001) [reply frontend/appflowy_flutter/packages/appflowy_backend/example/macos/Runner.xcodeproj/project.pbxproj:547]: sure.

**Comment by yatendra2001** [comment]
@appflowy wdyt is it ready to merge??

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart [hunk 7]
```diff
@@ -92,10 +94,26 @@ class SidebarUser extends StatelessWidget {
 
   Widget _buildUserName(BuildContext context, MenuUserState state) {
     final String name = _userName(state.userProfile);
-    return FlowyText.medium(
-      name,
-      overflow: TextOverflow.ellipsis,
-      color: Theme.of(context).colorScheme.tertiary,
+    return GestureDetector(
+      onTap: () async {
+        final result = await OpenAIService.requestTextCompletion(
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart:99]: I added this code in my previous PR as a sample code. You shouldn't include it here.

Remove the GestureDetector.

File: frontend/appflowy_flutter/packages/flowy_infra_ui/android/.project [hunk 8]
```diff
@@ -22,12 +22,12 @@
 	</natures>
 	<filteredResources>
 		<filter>
-			<id>1626576261667</id>
+			<id>1693395487121</id>
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/packages/flowy_infra_ui/android/.project:25]: Please revert the changes related to Android if they are not required.

File: frontend/appflowy_flutter/packages/flowy_infra_ui/android/.settings/org.eclipse.buildship.core.prefs [hunk 9]
```diff
@@ -1,2 +1,13 @@
...
+connection.gradle.distribution=GRADLE_DISTRIBUTION(VERSION(7.4.2))
 connection.project.dir=
 eclipse.preferences.version=1
+gradle.user.home=
+java.home=/Library/Java/JavaVirtualMachines/temurin-17.jdk/Contents/Home
+jvm.arguments=
+offline.mode=false
+override.workspace.settings=true
+show.console.view=true
+show.executions.view=true
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/packages/flowy_infra_ui/android/.settings/org.eclipse.buildship.core.prefs:13]: Please revert the changes related to Android if they are not required.

File: frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart (hunk 7 above)
Comment (yatendra2001) [reply frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/sidebar_user.dart:99]: sure.

File: frontend/appflowy_flutter/packages/flowy_infra_ui/android/.project (hunk 8 above)
Comment (yatendra2001) [reply frontend/appflowy_flutter/packages/flowy_infra_ui/android/.project:25]: done ✅

File: frontend/appflowy_flutter/packages/flowy_infra_ui/android/.settings/org.eclipse.buildship.core.prefs (hunk 9 above)
Comment (yatendra2001) [reply frontend/appflowy_flutter/packages/flowy_infra_ui/android/.settings/org.eclipse.buildship.core.prefs:13]: done ✅

**Comment by yatendra2001** [comment]
@LucasXu0 is it gtg??

**Comment by LucasXu0** [comment]
@yatendra2001 It appears that the changes related to Android have not been reverted.

Screenshot 2023-09-12 at 14 36 18

**Comment by yatendra2001** [comment]
> @yatendra2001 It appears that the changes related to Android have not been reverted.
> 
> Screenshot 2023-09-12 at 14 36 18

@LucasXu0 could you check this https://github.com/AppFlowy-IO/AppFlowy/pull/3302/commits/824671abf9ea96597ce64f558860d89a81739457

in this commit, i reverted those changes.

**Comment by yatendra2001** [comment]
@appflowy did we remove the part where flowy-ai was getting connected to flutter interface ??

Shall I write that again in next commit along with the interface??

**Comment by appflowy** [comment]
The flowy-ai is used in the backend, so it shouldn't expose any interfaces to Flutter. For instance, there could be a settings page that allows users to select which AI service to use, and then we can switch the service in the backend.

**Comment by yatendra2001** [comment]
> The flowy-ai is utilized in the backend, so it shouldn't expose any interfaces to Flutter. For instance, there could be a settings page that allows users to select the AI service they prefer, and we can subsequently adjust the service in the backend. I'm in the process of refactoring the AI, and it's still a work in progress.
> 
> After investigating further into the AI services within AppFlowy, I believe we should support a broader range of AI services, not just OpenAI. Therefore, we need to establish an abstraction layer, which necessitates refactoring the existing codebase, specifically the flowy-plugins.

This actually makes so much sense!!!

For sure we need an abstract layer of backend that shouldn't be exposed to frontend. But my doubt is then how are we going to make calls from the flutter interface ??

Also @appflowy , I was working on this project as part of #2782 , so how do i change the route or scope of the project? Let me know how can I kind of facilitate and advance this project in the direction we want at appflowy.
